File: dashboard/src/cnvm/compiler.ts
import { clampWeight } from "./compiler_types";
import type { CompiledCheckpoint, CompiledLayer, DomainRouting, RuleNir, VocabToken } from "./compiler_types";
import { slidersToVector } from "./tsr";

export { clampWeight };
export type { RuleNir, VocabToken, DomainRouting, CompiledLayer, CompiledCheckpoint };

export class CNVMCompiler {
  vocab_data: Record<string, VocabToken>;
  routing_data: Record<string, DomainRouting>;
  rules_data: Record<string, RuleNir>;
  tsrMap: Record<string, [number, number]>;
  dim: number;

  constructor(
    vocab_data: Record<string, VocabToken>,
    routing_data: Record<string, DomainRouting>,
    rules_data: Record<string, RuleNir>,
    tsrMap: Record<string, [number, number]>,
    dim: number
  ) {
    this.vocab_data = vocab_data;
    this.routing_data = routing_data;
    this.rules_data = rules_data;
    this.tsrMap = tsrMap;
    this.dim = dim;
  }

  private getRegisterOffset(name: string): number {
    if (!(name in this.tsrMap)) {
      throw new Error(`Unknown register name: ${name}`);
    }
    return this.tsrMap[name][0];
  }

  compileEmbedding(): { embedding: number[][]; token_to_id: Record<string, number> } {
    let maxId = 0;
    const token_to_id: Record<string, number> = {};
    
    for (const [token, data] of Object.entries(this.vocab_data)) {
      const tokenId = data.token_id;
      token_to_id[token] = tokenId;
      if (tokenId > maxId) {
        maxId = tokenId;
      }
    }

    const embedding: number[][] = Array.from(
      { length: maxId + 1 },
      () => new Array(this.dim).fill(0)
    );

    for (const [token, data] of Object.entries(this.vocab_data)) {
      const tokenId = token_to_id[token];
      const sliders: Record<string, number> = {};

      // Resolve ESC inheritance
      if (data.inherits) {
        const parentName = data.inherits;
        const parentData = this.vocab_data[parentName];
        if (parentData && parentData.sliders) {
          Object.assign(sliders, parentData.sliders);
        }
        if (data.overrides) {
          Object.assign(sliders, data.overrides);
        }
      } else {
        if (data.sliders) {
          Object.assign(sliders, data.sliders);
        }
      }

      // Clamp values and build vector
      const clampedSliders: Record<string, number> = {};
      for (const [k, v] of Object.entries(sliders)) {
        const numVal = typeof v === 'object' && v !== null && 'value' in v ? (v as any).value : v;
        clampedSliders[k] = clampWeight(Number(numVal));
      }

      embedding[tokenId] = slidersToVector(clampedSliders, this.tsrMap, this.dim);
    }

    return { embedding, token_to_id };
  }

  compileDrf(): { W_q: number[][]; W_k: number[][]; W_v: number[][] } {
    const W_q = Array.from({ length: this.dim }, () => new Array(this.dim).fill(0));
    const W_k = Array.from({ length: this.dim }, () => new Array(this.dim).fill(0));
    const W_v = Array.from({ length: this.dim }, () => new Array(this.dim).fill(0));

    // 1. Setup diagonal blocks (identity matrices) for domain isolation
    for (const [domain, [start, end]] of Object.entries(this.tsrMap)) {
      if (domain === "SYNTAX::POSITION_INDEX" || domain === "SYNTAX::POSITION_REL" || domain === "RESERVED") {
        continue;
      }
      for (let i = start; i <= end; i++) {
        W_q[i][i] = 1.0;
        W_k[i][i] = 1.0;
        if (i < 94) {
          W_v[i][i] = 1.0;
        }
      }
    }

    // 2. Punch bridges (SIDBs)
    for (const [domain, config] of Object.entries(this.routing_data)) {
      if (config.bridges) {
        try {
          const startIdx = this.getRegisterOffset(domain);
          for (const [targetDomain, weight] of Object.entries(config.bridges)) {
            const targetIdx = this.getRegisterOffset(targetDomain);
            const clampedW = clampWeight(weight);
            W_q[startIdx][targetIdx] = clampedW;
            W_k[startIdx][targetIdx] = clampedW;
            W_v[startIdx][targetIdx] = clampedW;
          }
        } catch {
          // Skip if domain doesn't exist in tsrMap to prevent crash during dynamic edits
        }
      }
    }

    // 3. Normalize Q and K matrices by baseline_d^(1/4) so logits product Q K^T is scaled by sqrt(baseline_d)
    const scalingFactor = Math.pow(94.0, 0.25);
    for (let i = 0; i < this.dim; i++) {
      for (let j = 0; j < this.dim; j++) {
        W_q[i][j] /= scalingFactor;
        W_k[i][j] /= scalingFactor;
      }
    }

    return { W_q, W_k, W_v };
  }

  compileSerg(): Record<number, CompiledLayer> {
    const rulesByLayer: Record<number, RuleNir[]> = {};
    for (const rule of Object.values(this.rules_data)) {
      const lIdx = rule.layer_index;
      if (!rulesByLayer[lIdx]) {
        rulesByLayer[lIdx] = [];
      }
      rulesByLayer[lIdx].push(rule);
    }

    const serg: Record<number, CompiledLayer> = {};

    for (const [lStr, rules] of Object.entries(rulesByLayer)) {
      const lIdx = parseInt(lStr, 10);
      const numRules = rules.length;

      const W_in = Array.from({ length: this.dim }, () => new Array(numRules).fill(0));
      const W_out = Array.from({ length: numRules }, () => new Array(this.dim).fill(0));
      const b = new Array(numRules).fill(-0.5); // Default bias is -0.5 (threshold)
      const ruleIds = rules.map(r => r.rule_id);

      for (let c = 0; c < numRules; c++) {
        const rule = rules[c];
        try {
          const triggerOffset = this.getRegisterOffset(rule.trigger_slider_name);
          const resultOffset = this.getRegisterOffset(rule.result_slider_name);
          const gateIn = clampWeight(rule.gate_in_weight);
          const gateOut = clampWeight(rule.gate_out_weight);

          // Preconditions
          W_in[triggerOffset][c] = gateIn;

          // Postconditions
          W_out[c][resultOffset] = gateOut;
          
          // Provenance Rule ID write-back to META::PROVENANCE (if it exists)
          if ("META::PROVENANCE" in this.tsrMap) {
            const provOffset = this.getRegisterOffset("META::PROVENANCE");
            W_out[c][provOffset] = rule.rule_id;
          }
        } catch {
          // Skip missing registers to handle dirty editing states gracefully
        }
      }

      serg[lIdx] = { W_in, W_out, b, rule_ids: ruleIds };
    }

    return serg;
  }

  compile(): CompiledCheckpoint {
    const { embedding, token_to_id } = this.compileEmbedding();
    const { W_q, W_k, W_v } = this.compileDrf();
    const serg = this.compileSerg();

    const id_to_token: Record<number, string> = {};
    for (const [k, v] of Object.entries(token_to_id)) {
      id_to_token[v] = k;
    }

    return {
      embedding,
      token_to_id,
      id_to_token,
      W_q,
      W_k,
      W_v,
      serg
    };
  }
}

File: cnvm/compiler.py
import numpy as np
from cnvm.tsr import get_dim, get_tsr_map, get_register_offset, sliders_to_vector


def clamp_weight(val: float, limit: float = 2.0) -> float:
    """Clamps a weight value to [-limit, limit] to prevent activation explosions."""
    return max(min(float(val), limit), -limit)


class CNVMCompiler:
    def __init__(self, vocab_data: dict, routing_data: dict, rules_data: dict, architecture_data: dict = None):
        self.vocab_data = vocab_data
        self.routing_data = routing_data
        self.rules_data = rules_data
        self.architecture_data = architecture_data or {
            "max_layers": 40,
            "cce_layers": [30, 31, 32, 33, 34],
            "default_cce_max_iter": 10,
            "default_cce_epsilon": 0.1
        }

    def compile_embedding(self) -> tuple[np.ndarray, dict]:
        """
        Compiles the vocabulary ontology with ESC inheritance handling.

        Returns:
            embedding_matrix: np.ndarray of shape (max_token_id + 1, DIM)
            token_to_id: dict mapping token string → integer ID
        """
        DIM = get_dim()
        max_id = 0
        token_to_id: dict[str, int] = {}

        for token, data in self.vocab_data.items():
            token_id = int(data["token_id"])
            token_to_id[token] = token_id
            if token_id > max_id:
                max_id = token_id

        # Zero-initialise the full embedding matrix
        embedding_matrix = np.zeros((max_id + 1, DIM), dtype=np.float32)
        
        # Set default value of safety facts to -2.0
        safety_facts = {
            "FACT::BOILER_ACTIVE",
            "FACT::WATER_LEVEL_LOW",
            "FACT::STEAM_PRESSURE_HIGH",
            "FACT::TEMPERATURE_SCALD",
            "FACT::SAFETY_VALVE_STUCK"
        }
        tsr_map = get_tsr_map()
        for name in safety_facts:
            if name in tsr_map:
                idx = tsr_map[name][0]
                embedding_matrix[:, idx] = -2.0

        system_defaults = {
            "SYS::CONFIDENCE": 2.0,
            "SYS::INTEGRITY": 2.0,
        }
        for name, val in system_defaults.items():
            if name in tsr_map:
                idx = tsr_map[name][0]
                embedding_matrix[:, idx] = val

        for token, data in self.vocab_data.items():
            token_id = token_to_id[token]
            sliders: dict[str, float] = {}

            # Resolve ESC Inheritance
            if "inherits" in data:
                parent_name = data["inherits"]
                parent_data = self.vocab_data.get(parent_name, {})
                if "sliders" in parent_data:
                    sliders.update(parent_data["sliders"])
                if "overrides" in data:
                    sliders.update(data["overrides"])
            else:
                if "sliders" in data:
                    sliders.update(data["sliders"])

            # Clamp all values before writing
            clamped = {k: clamp_weight(v) for k, v in sliders.items()}
            
            token_vector = embedding_matrix[token_id].copy()
            for name, value in clamped.items():
                if name in tsr_map:
                    idx = tsr_map[name][0]
                    token_vector[idx] = float(value)
            embedding_matrix[token_id] = token_vector

        return embedding_matrix, token_to_id

    def compile_drf(self) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
        """
        Compiles block-diagonal Domain Routing Fabric matrices (W_q, W_k, W_v).

        - Enforces block-diagonal identity for domain isolation.
        - Inserts Sparse Inter-Domain Bridges (SIDBs) from routing_data.
        - Normalises Q and K by d^(1/4) so Q K^T is pre-scaled by sqrt(d).

        Returns:
            W_q, W_k, W_v — each np.ndarray of shape (DIM, DIM)
        """
        DIM = get_dim()
        tsr_map = get_tsr_map()

        W_q = np.zeros((DIM, DIM), dtype=np.float32)
        W_k = np.zeros((DIM, DIM), dtype=np.float32)
        W_v = np.zeros((DIM, DIM), dtype=np.float32)

        # 1. Identity blocks for intra-domain routing isolation
        for domain, (start, end) in tsr_map.items():
            if domain in {"SYNTAX::POSITION_INDEX", "SYNTAX::POSITION_REL", "RESERVED", "META::PROVENANCE", "META::EVIDENCE", "SYS::CONFLICT"}:
                continue
            for i in range(start, end + 1):
                if i < 94 or i == 106:
                    W_q[i, i] = 1.0
                    W_k[i, i] = 1.0
                    W_v[i, i] = 1.0

        # 2. Compile Sparse Inter-Domain Bridges (SIDBs)
        for domain, config in self.routing_data.items():
            if "bridges" not in config:
                continue
            start_idx = get_register_offset(domain)
            for target_domain, weight in config["bridges"].items():
                target_idx = get_register_offset(target_domain)
                clamped_w = clamp_weight(weight)
                W_q[start_idx, target_idx] = clamped_w
                W_k[start_idx, target_idx] = clamped_w
                W_v[start_idx, target_idx] = clamped_w

        # 3. Normalise: divide Q and K by baseline_d^0.25 so Q K^T / sqrt(baseline_d) is stable
        scaling_factor = 94.0 ** 0.25
        W_q_norm = W_q / scaling_factor
        W_k_norm = W_k / scaling_factor

        return W_q_norm, W_k_norm, W_v

    def compile_serg(self) -> dict[int, dict]:
        """
        Compiles rules_data into Sparse Executable Rule Graph matrices, grouped by layer.

        Each entry in the returned dict maps layer_index → {W_in, W_out, b, rule_ids}.
          W_in:  (DIM, R)  — precondition key detectors
          W_out: (R, DIM)  — postcondition effect + provenance writers
          b:     (R,)      — threshold bias (default -0.5)
        """
        DIM = get_dim()

        # Group rules by layer_index
        rules_by_layer: dict[int, list] = {}
        for rule_name, rule_info in self.rules_data.items():
            l_idx = int(rule_info["layer_index"])
            rules_by_layer.setdefault(l_idx, []).append(rule_info)

        compiled_layers: dict[int, dict] = {}
        for l_idx, rules in rules_by_layer.items():
            R = len(rules)
            W_in  = np.zeros((DIM, R), dtype=np.float32)
            W_out = np.zeros((R, DIM), dtype=np.float32)
            b     = np.full(R, -0.5, dtype=np.float32)   # threshold bias per paper §8.1

            for c, rule in enumerate(rules):
                trigger_offset = get_register_offset(rule["trigger_slider_name"])
                result_offset  = get_register_offset(rule["result_slider_name"])
                gate_in  = clamp_weight(rule["gate_in_weight"])
                gate_out = clamp_weight(rule["gate_out_weight"])
                rule_id  = int(rule["rule_id"])

                # Precondition key detector
                W_in[trigger_offset, c] = gate_in

                # Postcondition: effect on result register
                W_out[c, result_offset] = gate_out

                # Provenance: write rule_id into PROVENANCE_ID register
                prov_offset = get_register_offset("META::PROVENANCE")
                W_out[c, prov_offset] = 0.0

            compiled_layers[l_idx] = {
                "W_in": W_in,
                "W_out": W_out,
                "b": b,
                "rule_ids": [r["rule_id"] for r in rules],
            }

        return compiled_layers

    def compile(self) -> dict:
        """Compiles the full CNVM tensor checkpoint from all NIR namespaces."""
        embedding_matrix, token_to_id = self.compile_embedding()
        W_q, W_k, W_v = self.compile_drf()
        serg_layers = self.compile_serg()

        return {
            "embedding":   embedding_matrix,
            "token_to_id": token_to_id,
            "id_to_token": {v: k for k, v in token_to_id.items()},
            "W_q":  W_q,
            "W_k":  W_k,
            "W_v":  W_v,
            "serg": serg_layers,
            "architecture": self.architecture_data,
        }
